docs/specs/oidc-native-sso.md

```diff
@@ -0,0 +1,153 @@
+- [OIDC Native SSO](#oidc-native-sso)
+  * [Key points to note](#key-points-to-note)
+  * [Changes on Client](#changes-on-client)
+  * [Changes on OfflineGrant](#changes-on-offlinegrant)
+  * [Changes on Authorization Endpoint](#changes-on-authorization-endpoint)
+  * [Changes on Token Endpoint](#changes-on-token-endpoint)
+    + [`grant_type=authorization_code`](#grant_typeauthorization_code)
+    + [`grant_type=urn:authgear:params:oauth:grant-type:biometric-request`](#grant_typeurnauthgearparamsoauthgrant-typebiometric-request)
+    + [`grant_type=refresh_token`](#grant_typerefresh_token)
+    + [`grant_type=urn:authgear:params:oauth:grant-type:app2app`](#grant_typeurnauthgearparamsoauthgrant-typeapp2app)
+    + [`grant_type=urn:ietf:params:oauth:grant-type:token-exchange`](#grant_typeurnietfparamsoauthgrant-typetoken-exchange)
+    + [When issuing ID tokens](#when-issuing-id-tokens)
+  * [Changes on Admin API](#changes-on-admin-api)
+  * [Changes on SDK](#changes-on-sdk)
+    + [Recipe: Two applications written by the same vendor](#recipe-two-applications-written-by-the-same-vendor)
+    + [Recipe: A application opening webapps with custom webview](#recipe-a-application-opening-webapps-with-custom-webview)
+  * [Caveats](#caveats)
+
+# OIDC Native SSO
+
+This document specifies the implementation of [OIDC Native SSO](https://openid.net/specs/openid-connect-native-sso-1_0.html).
+
+## Key points to note
+
+- A OfflineGrant used to has only one set of client-specific information, like `client_id` and `refresh_token`. Now, a OfflineGrant can have multiple refresh tokens.
+- Since there is only one OfflineGrant, the apps sharing user authentication with Native SSO does not share refresh token. But the underlying session is shared. Thus signing out in one app will sign out all apps. This is by design. See [App2app](./app2app.md) if you want the apps have independent sessions.
+- Native SSO is done through the Token Endpoint, thus it requires no user interaction.
+- Existing OfflineGrant cannot perform Native SSO. It is because `device_sso` is not in the `scope`. Sign in again to obtain a Native SSO OfflineGrant to perform Native SSO.
+
+## Changes on Client
+
+- Add `x_device_sso_enabled: boolean`.
+- `scope=device_sso` is allowed if `x_device_sso_enabled=true`.
+
+> Do we need to add `x_device_sso_key: string` to designate which group of clients can perform Native SSO?
+> Only clients with `x_device_sso_enabled=true` AND the same value of `x_device_sso_key` can perform Native SSO with each other.
+> This seems very advanced to me.
+
+## Changes on OfflineGrant
+
+- The following fields become client-specific
+  - `ClientID`
+  - `AuthorizationID`
+  - `IdentityID`
+  - `Scopes`
+  - `TokenHash`
+- Add a new field `DeviceSSORefreshTokens`. It has the above client-specific fields.
+  - `ClientID`
+  - `AuthorizationID`
+  - `IdentityID`
```

Comment:
It seems `AuthorizationID` and `IdentityID` does not need to be token specific:

- AuthorizationID points to an Authorization. And Authorization actually refers to a "group" of session displayed in the settings page. When you terminate a session, you are actually deleting an Authorization. In native SSO, the new refresh tokens are actually sharing a same session with the original session, thus the Authorization should be the same. It represents the "parent" offline grant.
- IdentityID is for biometric login. As biometric login always creates new offline grant, it won't exist in the child refresh tokens?

Reply:
Authorization associates a Client with a User. The refresh token is now client-specific. Actually a session can now be shared by different clients.

---

Your reasoning on IdentityID seems correct. Would there be any problem if we still make IdentityID token specific?

Reply:
> Authorization associates a Client with a User. The refresh token is now client-specific. Actually a session can now be shared by different clients.

Yes, now session can be shared by different client, but that is already represented by the `ClientID` of each token. There should be a "initial" client, which seems represented by "Authorization"? Conceptually, what actually is "Authorization"? 

> Your reasoning on IdentityID seems correct. Would there be any problem if we still make IdentityID token specific?

No problem, but IdentityID will always be empty.

Reply:
- Each token will have different Authorization which is unique per client id.
- Update scope of existing Authorization if already exist and if needed.